maro/simulator/scenarios/vm_scheduling/business_engine.py

```diff
@@ -0,0 +1,492 @@
+            )
+            vm_request_event = self._event_buffer.gen_cascade_event(
+                tick=tick,
+                event_type=Events.REQUEST,
+                payload=vm_req_payload
+            )
+            self._event_buffer.insert_event(event=vm_request_event)
+            self._total_vm_requests += 1
+
+    def post_step(self, tick: int):
+        # Update energy to the environment metrices.
+        total_energy: float = 0.0
+        for pm in self._machines:
+            total_energy += pm.energy_consumption
+        self._total_energy_consumption += total_energy
+
+        if (tick + 1) % self._snapshot_resolution == 0:
+            # NOTE: We should use frame_index method to get correct index in snapshot list.
+            self._frame.take_snapshot(self.frame_index(tick))
+
+        # Stop current episode if we reach max tick.
+        return tick + 1 >= self._max_tick
+
+    def get_event_payload_detail(self) -> dict:
+        """dict: Event payload details of current scenario."""
+        return {
+            Events.REQUEST.name: VmRequestPayload.summary_key,
+            MaroEvents.PENDING_DECISION.name: DecisionPayload.summary_key
+        }
+
+    def get_agent_idx_list(self) -> List[int]:
+        """Get a list of agent index.
+
+        Returns:
+            list: List of agent index.
+        """
+        return [0]
+
+    def get_node_mapping(self) -> dict:
+        """dict: Node mapping."""
+        node_mapping = {}
+
+        return node_mapping
+
+    def get_metrics(self) -> DocableDict:
+        """Get current environment metrics information.
+
+        Returns:
+            DocableDict: Metrics information.
+        """
+
+        return DocableDict(
+            metrics_desc,
+            total_vm_requests=self._total_vm_requests,
+            total_energy_consumption=self._total_energy_consumption,
+            successful_placement=self._successful_placement,
+            successful_completion=self._successful_completion,
+            failed_placement=self._failed_placement,
+            total_latency=self._total_latency,
+            total_oversubscriptions=self._total_oversubscriptions
+        )
+
+    def _register_events(self):
+        # Register our own events and their callback handlers.
+        self._event_buffer.register_event_handler(event_type=Events.REQUEST, handler=self._on_vm_required)
+        # Generate decision event.
+        self._event_buffer.register_event_handler(event_type=MaroEvents.TAKE_ACTION, handler=self._on_action_received)
+
+    def _update_vm_workload(self, cur_tick_cpu_utilization: dict):
+        """Update all live VMs CPU utilization.
+
+        The length of VMs utilization series could be difference among all VMs,
+        because index 0 represents the VM's CPU utilization at the tick it starts.
+        """
+        for live_vm in self._live_vms.values():
+            # NOTE: Some data could be lost. We use -1.0 to represent the missing data.
```

Comment:
I think we can apple the back padding in the data processing.

Reply:
@Jinyu-W

Reply:
Currently we use the utilization of last tick to pad this.

We think we can also use the MA of the recent fixed window size or some other methods. So we implement it here to avoid pre-process data again and again.

If we choose only one way to pad the lost data, I agree done in the preprocessing part is better.

Which one do you prefer?